### ISSUE TITLE
fflush() regression

### ISSUE THREAD

**Issue by juj** [emscripten-core/emscripten#20622]
`a.cpp`

```c++
#include <stdio.h>

int main () {
    printf("ok");
    fflush(stdout);
}
```

Results in

![image](https://github.com/emscripten-core/emscripten/assets/225351/b3f98727-1043-4d5a-9bcd-fc7d181c5f55)

should have printed `ok`, but instead gave a warning about not having fflushed.

**Comment by juj** [comment]
Actually, nevermind, I realize I had alocal change of having flipped EXIT_RUNTIME around, and the behavior is exactly like the warning there states.

**Comment by juj** [comment]
Err, closed in confusion. It seems that in the above code, `fflush()` does nothing, but it should have flushed.

**Comment by juj** [comment]
Passing `-sFORCE_FILESYSTEM=1` does not seem to help.

**Comment by RReverser** [comment]
I don't think this ever worked as expected? IIRC there were issues about using `process.stdout.write` instead of `console.log` so that it could write partial lines at least in Node.js but I don't think it was implemented.

**Comment by RReverser** [comment]
I was probably thinking of #17320.

**Comment by juj** [comment]
> I don't think this ever worked as expected?
Yeah, I agree. `fflush(stdout)` seems to have been a no-op since dawn of time.

The root clash with `fflush` is that `console.log()` implements line-based printing, so if one had
```c++
int main()
{
  printf("hel");
  fflush(stdout);
  printf("lo");
  fflush(stdout);
}
```
Then if `fflush` is implemented, it would need to introduce a newline to print `hel\nlo\n`. Many users might have unbuffered input in the form that would instantaneously flush after each printed char or token - and making `fflush` introduce a newline (even if in a smart way, only if flushed content didn't already end in a natural endline) would cause broken printing for such users.

(I recall working on a branch some near decade ago that changed printing to be character-based, and line buffering and flushing would be implemented in the `Module.print()` function - that way if users were printing on a DOM element, they wouldn't have this problem)

---------

Upon further debugging, it looks like the root regression that I am seeing is somewhere with `--emrun` integration instead.

Whereas 
`a.cpp`
```c++
#include <stdio.h>
int main () { printf("ok"); }
```
with
```
em++ a.cpp --emrun -sEXIT_RUNTIME=1 -o a.html && node a.js
```
prints as expected, it looks like
```
em++ a.cpp --emrun -sEXIT_RUNTIME=1 -o a.html && emrun a.html
```
no longer prints `ok`, but gives

![image](https://github.com/emscripten-core/emscripten/assets/225351/99035f68-d48a-4e5b-a991-d040b500d1be)

This used to work in the past. I'll need to dig in a bit further.

**Comment by kripken** [comment]
I am also pretty sure this worked in the past - `fflush` should flush both the musl buffer and the JS buffer, and get everything printed out. So this does look like a regression. Bisection might be the fastest way to figure this out.

**Comment by sbc100** [comment]
But how can you "flush" a string without a newline in it on the web where you only have `console`?    How can that have ever worked?

**Comment by kripken** [comment]
It should force a newline on the JS side, I think. But now that you mention it, maybe it intentionally doesn't? If so maybe my and @juj's memory are wrong and it never worked?

**Comment by sbc100** [comment]
I think the above program should never produce "hel" and "lo" on two different lines.. that would just be wrong.
